README
                        README FILE FOR SPIM, XSPIM, PCSPIM, and QTSPIM
                        ===============================================

This directory contains SPIM--an assembly language MIPS32 simulator.

SPIM is Copyright (c) 1990-2015, by James R. Larus.
All rights reserved.

SPIM is distributed under a BSD license:

     Redistribution and use in source and binary forms, with or without modification, are
     permitted provided that the following conditions are met:

         Redistributions of source code must retain the above copyright notice, this list of
         conditions and the following disclaimer.

         Redistributions in binary form must reproduce the above copyright notice, this list of
         conditions and the following disclaimer in the documentation and/or other materials
         provided with the distribution.

         Neither the name of the James R. Larus nor the names of its contributors may be used to
         endorse or promote products derived from this software without specific prior written
         permission.

     THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS
     OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
     MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
     COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
     EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
     SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
     HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
     TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
     SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


QtSPIM is linked to the Qt library, which is distributed under the GNU Lesser General Public License
version 3 and GNU Lesser General Public License version 2.1.

GNU Lesser General Public License, version 2.1  http://www.gnu.org/licenses/old-licenses/lgpl-2.1.html
GNU Lesser General Public License, version 3  http://www.gnu.org/licenses/lgpl-3.0.html


                                      GETTING SPIM SOURCE
                                      ===================

Spim is hosted on SourceForge and its source code is available through SVN:
https://sourceforge.net/projects/spimsimulator/


                                          DIRECTORIES
                                          ===========

The subdirectories can be divided into two categories.

1. Common to all versions of Spim:

    CPU -- Contains the MIPS simulator and some utility functions common to all version of the
           simulator.

    Documentation -- Contains the man pages, web page, and other documentation.

    bin -- Utilities.

    Setup -- Contain tools used to build deployable versions of the simulators.

    Tests -- Contains regression test for the simulator.


2. Front-ends for different versions of Spim:

    PCSpim -- Microsoft Windows front end for Spim. [Obsolete]

    QtSpim -- Front end for Spim that works on Microsoft Windows, Linux, and Mac OS X.

    spim -- Text-only (terminal or console) front end for Spim. Works on all platforms.

    xspim -- X-Windows front end for Spim. Works on Linux and Mac OS X. (Built on a very old
             X Windows library that is no longer supported by the Mac.) [Obsolete]


                                        BUILDING QTSPIM
                                        ===============

It is generally not necessary to build QtSpim as the Spim website contains an installable
version for most systems.

Install the bison parser generator and flex lexer generator, either using Cygwin on
Windows or native version on Linux and OS X.

1. Download the free (open source) version of Qt Creator from http://qt.io. This is a
large download that might take a while.  I have been using version 5.5, though any of the
version 5.x should work fine. The 4.x releases will cause problems because of incompatible
changes to the libraries.

        The open source version of Qt Creator for Windows, which uses mingw, works fine,
        as does the version that uses Microsoft Visual Studio's C++ compiler.

        The Qt Creator package for Ubuntu is typically several releases behind, which can
        cause some problems with QtSPIM.  Download a recent version from http://qt.io.

2. Start QtCreator and configure the project to use whichever compilers are available on
your system. gcc, llvm, and mdvc all work, though some of them generate some warnings at
high optimization levels (yes, I should fix them).

3. Build QtSpim.

        If you are porting to a new system and see a large number of compiler errors in
        QtCreator, changes are good that you need to change QtSpim.pro to define the
        compiler flag that says "treat all files, including .c files, as C++ code. This
        flag is "-X C++" for gcc and "-TP" for the Microsoft compiler.

4. Packaging QtSpim for distribution is a complex, OS-specific operation. The Setup
directory contains the files for all three platforms.

	On OS X, you need to run the script bin/release-osx and use the free Packages tool
	(http://s.sudre.free.fr/Software/Packages/about.html).

	On Windows,you need a copy of Visual Studio and the free version of InstallShield.

	On Linux, the script bin/release-debian handles the entire process.

On all three systems, the packaged version of QtSpim includes all of the libraries and
plugins used by QtSpim and assistant (the Qt help file viewer). Previously, the Linux used
dependencies to bring in other packages that contained these libraries, but I cannot find
the packages that contain all of these files for Qt5, so it is easier just to redistribute
everything.


                                    BUILDING SPIM and XSPIM
                                    =======================

It is necessary to build spim and xspim, as they are only distributed as source.

   1. Download the source for spimsimulator from sourceforge
   (https://sourceforge.net/projects/spimsimulator/), either through Subversion from the
   spimsimulator repository (http://spimsimulator.sourceforge.net) or
   by downloading the source code (https://sourceforge.net/projects/spimsimulator/files).

   2. Decompress the file, using either the program uncompress for the first file or gzip for
      the second file:

          % uncompress spim.tar.Z

      or

          % gzip -d spim.tar.gz

   3. Move the file spim.tar to the directory in which you want to build spim and untar it:

          % tar xf spim.tar

      It will create a directory named spim-8.0 (or the most recent version number).

   4. The simple terminal interface is contained in the spim-8.0/spim directory and the
   X-windows interfaces is in the spim-8.0/xspim directory. The other directories are described
   in the README file.

   5. Next, you must set the directories in which spim will be installed by editing the Makefile
   (the file that contains instructions on building spim). In general, if you are installing
   spim and want the windowing version (xspim), edit the file xspim/Imakefile. If you don't want
   xspim or are running on a system without X-windows installed, you use the file spim/Makefile.

      The programs are installed in standard locations, but you can change the pathnames to other locations:

          EXCEPTION_DIR -- The full pathname of the directory in which to install the spim exception handler (exceptions.s).

          BIN_DIR -- The full pathname of the directory in which spim and xspim should be installed.

          MAN_DIR -- The full pathname of the directory in which the manual pages for spim and xspim should be installed.

      In general, the remaining parameters in a Makefile need not be changed.

   6. Then, if you are building xspim, change to the spim-8.0/xspim directory and type:

          % xmkmf
          % make

      If you do not have a copy of xmkmf, you can use the Makefile in the xspim directory, but
      beware that it may not work on your system because the paths to the X windows libraries
      could be different.

   7. If you do not have X-windows, change to thespim-8.0/spim directory, edit Makefile, and type:

          % make

   8. To run spim or xspim, the exception handler must be installed in the directory specified
   by the variable EXCEPTION_DIR in the Makefile. If the file exception.s is not installed, spim
   and xspim fail before they start running. You can either install this file by hand or by
   typing:

          % make install

   which also installs spim or xspim, and the manual pages in the directories that you set
   (above). You may need root permission to install these files, in which case type:

          % sudo make install

   9. To test that spim is correctly built, change to the spim-8.0/spim directory and type:

          % make test

      and examine the output of the test. (Note: the exception handler must be installed before
      running the test.)


                                        BUILDING PCSPIM
                                        ===============

It is not necessary to compile PCSpim (the Microsoft Windows version), as the Spim web site
contains a precompiled version.

The Microsoft Windows version of Spim is called PCSpim and is built using Microsoft Visual
Studio.

spim (the terminal version) works well on Microsoft Windows.  You can easily build it using the
Cygwin port of the GNU tools (see www.cygwin.com).



                                      NEW VERSIONS OF SPIM
                                      ===================

I generally release new version of SPIM once a year, before a semester boundary (late August or
early January).  The new version are available on:

        https://sourceforge.net/projects/spimsimulator/


ChangeLog
2018-11-23  Larus  <larus@Jim-Macbook-Air>

	* README (https):
	Fix URLs for sourceforge.

2018-11-09  Larus  <larus@Jim-Macbook-Air>

	* Documentation/H_P_AppendixA/HP_AppA.html [https://sourceforge.net/p/spimsimulator/bugs/86/]:
	Opcode for swc1 was documented incorrectly as 0x31 instead of 0x39.

2018-09-16  Larus  <larus@Jim-Macbook-Air>

	* spim/Makefile, xspim/Makefile
	[https://sourceforge.net/p/spimsimulator/patches/6]:
	Improvements to Makefiles

2017-08-29  Larus  <larus@Jim-Macbook-Air>

	* CPU/version.h (SPIM_VERSION):
	Release 9.1.20

	* spimview.ui [https://sourceforge.net/p/spimsimulator/bugs/81]:
	Typo "parmeter"

	* state.cpp [https://sourceforge.net/p/spimsimulator/bugs/81]:
	Add message warning QtSpim will exit to restore window state.


2017-07-28    <larus@IC-LMC-ICDO-01>

	* Setup/QtSpim_Win_Deployment:
	Delete VisualStudio projects (InstallShield no longer works) and
	move to WiX to build installer.

	* QtSpim/menu.cpp (help_ViewHelp):
	Use full path for help file on Windows so QtSpim can run anywhere.

2017-07-25  Larus  <larus@Jim-Macbook-Air>

        * CPU/version.h (SPIM_VERSION):
        Release 9.1.19

2017-06-09  Larus  <larus@Jim-Macbook-Air>

        * QtSpim/main.cpp (main):
        Explicitly specify organization name and url to work around
        incompatibility on MacOS for settings.

2017-02-18  Larus  <larus@Jim-Macbook-Air>

        * QtSpim/regwin.cpp (formatFloat) [https://sourceforge.net/p/spimsimulator/bugs/76/]:
        Display binary values for floats to 32 bits.

2017-01-20  Larus  <larus@Jim-Macbook-Air>

        *
        Rebuild QtSpim on Mac with Qt5.5 because Qt5.7 has broken
        macdeployqt that does not update paths to frameworks in frameworks
        copied into the app.

2017-01-03  Larus  <larus@Jim-Macbook-Air>

        *
        Release 9.1.18

2017-01-03  James Larus  <larus@ubuntu>

        * control.prototype, qtspim.desktop:
        Update version number.

        * QtSpim/QtSpim.pro, bin/release-debian:
        Set RPATH in linker rather than using chrpath (which breaks in Ubuntu 16.04).

2017-01-03  Larus  <larus@Jim-Macbook-Air>

        * spimview.cpp:
        Update organization name.

        * macinfo.plist:
        Update version number.

        * spim-utils.cpp, version.h:
        Update copyright date and release date.

2016-11-25  Larus  <larus@Jim-Macbook-Air>

        * QtSpim/spimview.cpp, menu.cpp, settings.ui [Renaud PACALET pacalet@users.sf.ne]:
        Add button to reset exception handler to default.

        * QtSpim/QtSpim.pro, CPU/*.c, spim/Makefile:
        Move over completely to c++ compilation since the new version of
        Qt requires c++11.

2016-11-23  Larus  <larus@Jim-Macbook-Air>

        * CPU/version.h
        Version 9.1.18

        * Documentation/H_P_AppendixA/HHP_AppA.html:
        Fix register usage for syscalls in Table 9.1.

2016-02-27  Larus  <larus@Jim-Macbook-Air>

        *  QtSpim/menu.cpp:
        Unnecessary refresh of register display.

        * QtSpim/spimview.h, regwin.cpp:
        Did not highlight changes to double precision registers correctly.

2015-12-29  Larus  <larus@Jim-Macbook-Air>

        * CPU/version.h (SPIM_VERSION):
        Version 9.1.17

        * summary.rtf:
        Update installation instructions for latest OS X.

        * QtSpim/QtSpim.pro, menu.cpp [https://sourceforge.net/p/spimsimulator/bugs/71/]:
        Fix creation of qtspim.qhc file to eliminate embedded paths.

        *  menu.cpp, spimview.cpp, spimview.h:
        Highlight spim notice with color.

2015-12-28  Larus  <larus@ubuntu>

    * bin/release-debian [https://sourceforge.net/p/spimsimulator/support-requests/9/]:
    Forgot sqldriver on Linux.


2015-12-28  Larus  <larus@Jim-Macbook-Air>

        * CPU/scanner.l [https://sourceforge.net/p/spimsimulator/bugs/66/]:
        Improvement to previous fix for flex version detection.

2015-12-21  Larus  <larus@Jim-Macbook-Air>

        * QtSpim/state.cpp (writeSettings):
        Added check box for spim quiet mode.

2015-09-17    <larus@IC-LMC-ICDO-01>

        * QtSpim/regwin.cpp, QtSpim/datawin.cpp [https://sourceforge.net/p/spimsimulator/bugs/60/]
        Fix of bug 60 did not work on Windows because of difference in
        treatment of hex literals from Mac.

2015-08-27  Larus  <larus@Jim-Macbook-Air>

        * README, README.rtf;
        Update build instructions further.

2015-08-25  Larus  <larus@ubuntu>

        * CPU/version.h (SPIM_VERSION):
        Version 9.1.16

        * README, README.rtf:
        Update build instructions.

        * QtSpim/spimview.cpp (initialize_world):
        Update copyright notice for QtSpim to notify that Qt uses LGPL.

        * QtSpim/QtSpim.pro:
        Configuration name changed from linux-g++ -> linux-g++-32

2015-04-26  Larus  <larus@Jim-MacBook-Air.local>

        * CPU/version.h (SPIM_VERSION):
        Fix installation on Mac to include all libraries.
        Version 9.1.15.

2015-03-12  Larus  <larus@icdhcp-1-044.epfl.ch>

        * QtSpim/state.cpp (win_Restore):
        Add command on Windows menu to restore windows to initial default configuration.

2015-03-11  Larus  <larus@icdhcp-1-044.epfl.ch>

        * QtSpim/datawin.cpp (formatPartialQuadWord) [https://sourceforge.net/p/spimsimulator/bugs/61/]:
        Problem printing partial word at end of memory space.

        * CPU/scanner.l:
        Change of 2/18 made a number followed by a comma and space
        ambiguous. Now FP numbers must have a digit following decimal
        point (decimal comma).

2015-02-18  Larus  <larus@ubuntu>

        * QtSpim/regwin.cpp, QtSpim/datawin.cpp [https://sourceforge.net/p/spimsimulator/bugs/60/]
        Setting register and memory contents in QtSpim did not allow
        negative hex numbers (0xf....).


CPU/reg.h
/* SPIM S20 MIPS simulator.
   Declarations of registers and code for accessing them.

   Copyright (c) 1990-2010, James R. Larus.
   All rights reserved.

   Redistribution and use in source and binary forms, with or without modification,
   are permitted provided that the following conditions are met:

   Redistributions of source code must retain the above copyright notice,
   this list of conditions and the following disclaimer.

   Redistributions in binary form must reproduce the above copyright notice,
   this list of conditions and the following disclaimer in the documentation and/or
   other materials provided with the distribution.

   Neither the name of the James R. Larus nor the names of its contributors may be
   used to endorse or promote products derived from this software without specific
   prior written permission.

   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
   AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
   IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
   ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
   LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
   CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
   GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
   HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
   LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
   OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/


typedef int32 /*@alt unsigned int @*/ reg_word;
typedef uint32 u_reg_word;


/* General purpose registers: */

#define R_LENGTH	32

extern reg_word R[R_LENGTH];

extern reg_word HI, LO;

extern mem_addr PC, nPC;


/* Argument passing registers */

#define REG_V0		2
#define REG_A0		4
#define REG_A1		5
#define REG_A2		6
#define REG_A3		7
#define REG_FA0		12
#define REG_SP		29


/* Result registers */

#define REG_RES		2
#define REG_FRES	0


/* $gp registers */

#define REG_GP		28



/* Coprocessor registers: */

extern reg_word CCR[4][32], CPR[4][32];



/* Exeception handling registers (Coprocessor 0): */

/* BadVAddr register: */
#define CP0_BadVAddr_Reg 8
#define CP0_BadVAddr	(CPR[0][CP0_BadVAddr_Reg])

/* Count register: */
#define CP0_Count_Reg	9
#define CP0_Count	(CPR[0][CP0_Count_Reg]) /* ToDo */

/* Compare register: */
#define CP0_Compare_Reg	11
#define CP0_Compare	(CPR[0][CP0_Compare_Reg]) /* ToDo */

/* Status register: */
#define CP0_Status_Reg	12
#define CP0_Status	(CPR[0][CP0_Status_Reg])
/* Implemented fields: */
#define CP0_Status_CU	0xf0000000
#define CP0_Status_IM	0x0000ff00
#define CP0_Status_IM7  0x00008000 /* HW Int 5 */
#define CP0_Status_IM6  0x00004000 /* HW Int 4 */
#define CP0_Status_IM5  0x00002000 /* HW Int 3 */
#define CP0_Status_IM4  0x00001000 /* HW Int 2 */
#define CP0_Status_IM3  0x00000800 /* HW Int 1 */
#define CP0_Status_IM2  0x00000400 /* HW Int 0 */
#define CP0_Status_IM1  0x00000200 /* SW Int 1 */
#define CP0_Status_IM0  0x00000100 /* SW Int 0 */
#define CP0_Status_UM	0x00000010
#define CP0_Status_EXL	0x00000002
#define CP0_Status_IE	0x00000001
#define CP0_Status_Mask (CP0_Status_CU		\
			 | CP0_Status_UM	\
			 | CP0_Status_IM	\
			 | CP0_Status_EXL	\
			 | CP0_Status_IE)

/* Cause register: */
#define CP0_Cause_Reg	13
#define CP0_Cause	(CPR[0][CP0_Cause_Reg])
/* Implemented fields: */
#define CP0_Cause_BD	0x80000000
#define CP0_Cause_IP	0x0000ff00
#define CP0_Cause_IP7   0x00008000 /* HW Int 5 */
#define CP0_Cause_IP6   0x00004000 /* HW Int 4 */
#define CP0_Cause_IP5   0x00002000 /* HW Int 3 */
#define CP0_Cause_IP4   0x00001000 /* HW Int 2 */
#define CP0_Cause_IP3   0x00000800 /* HW Int 1 */
#define CP0_Cause_IP2   0x00000400 /* HW Int 0 */
#define CP0_Cause_IP1   0x00000200 /* SW Int 1 */
#define CP0_Cause_IP0   0x00000100 /* SW Int 0 */
#define CP0_Cause_ExcCode 0x0000007c
#define CP0_Cause_Mask	(CP0_Cause_BD		\
			 | CP0_Cause_IP		\
			 | CP0_Cause_IP7	\
			 | CP0_Cause_IP6	\
			 | CP0_Cause_IP5	\
			 | CP0_Cause_IP4	\
			 | CP0_Cause_IP3	\
			 | CP0_Cause_IP2	\
			 | CP0_Cause_ExcCode)
#define CP0_ExCode	((CP0_Cause & CP0_Cause_ExcCode) >> 2)

/* EPC register: */
#define CP0_EPC_Reg	14
#define CP0_EPC		(CPR[0][CP0_EPC_Reg])

/* Config register: */
#define CP0_Config_Reg	16
#define CP0_Config	(CPR[0][CP0_Config_Reg])
/* Implemented fields: */
#define CP0_Config_BE	0x000080000
#define CP0_Config_AT	0x000060000
#define CP0_Config_AR	0x00001c000
#define CP0_Config_MT	0x000000380
#define CP0_Config_Mask (CP0_Config_BE		\
			 | CP0_Config_AT	\
			 | CP0_Config_AR	\
			 | CP0_Config_MT)



/* Floating Point Coprocessor (1) registers.

   This is the MIPS32, Revision 1 FPU register set. It contains 32, 32-bit
   registers (either 32 single or 16 double precision), as in the R2010.
   The MIPS32, Revision 2 or MIPS64 register set has 32 of each type of
   register. */

#define FGR_LENGTH	32
#define FPR_LENGTH	16

extern double *FPR;		/* Dynamically allocate so overlay */
extern float *FGR;		/* is possible */
extern int *FWR;		/* is possible */


#define FPR_S(REGNO)	(FGR[REGNO])

#define FPR_D(REGNO)	(((REGNO) & 0x1) \
			 ? (run_error ("Odd FP double register number\n") , 0.0) \
			 : FPR[(REGNO) / 2])

#define FPR_W(REGNO)	(FWR[REGNO])


#define SET_FPR_S(REGNO, VALUE)	{FGR[REGNO] = (float) (VALUE);}

#define SET_FPR_D(REGNO, VALUE) {if ((REGNO) & 0x1) \
				 run_error ("Odd FP double register number\n"); \
				 else FPR[(REGNO) / 2] = (double) (VALUE);}

#define SET_FPR_W(REGNO, VALUE) {FWR[REGNO] = (int32) (VALUE);}


/* Floating point control registers: */

#define FCR		(CPR[1])


#define FIR_REG		0
#define FIR		(FCR[FIR_REG])
/* Implemented fields: */
#define FIR_W		0x0008000
#define FIR_D		0x0001000
#define FIR_S		0x0000800
#define FIR_MASK	(FIR_W | FIR_D | FIR_S)

#define FCCR_REG	25
#define FCCR		(FCR[FCCR_REG])
/* Implemented fields: */
#define FCCR_FCC	0x000000ff
#define FCCR_MASK	(FCCR_FCC)

#define FEXR_REG	26
#define FEXR		(FCR[FEXR_REG])
/* No implemented fields */

#define FENR_REG	28
#define FENR		(FCR[FENR_REG])
/* No implemented fields */

#define FCSR_REG	31
#define FCSR		(FCR[FCSR_REG])
/* Implemented fields: */
#define FCSR_FCC	0xfe800000
#define FCSR_MASK	(FCSR_FCC)
/* Floating point Cause (not implemented): */
#define FCSR_Cause_E	0x00020000
#define FCSR_Cause_V	0x00010000
#define FCSR_Cause_Z	0x00008000
#define FCSR_Cause_O	0x00004000
#define FCSR_Cause_U	0x00002000
#define FCSR_Cause_I	0x00001000
/* Floating point Enables (not implemented): */
#define FCSR_Enable_V	0x00000800
#define FCSR_Enable_Z	0x00000400
#define FCSR_Enable_O	0x00000200
#define FCSR_Enable_U	0x00000100
#define FCSR_Enable_I	0x00000080
/* Floating point Flags (not implemented): */
#define FCSR_Flag_V	0x00000040
#define FCSR_Flag_Z	0x00000020
#define FCSR_Flag_O	0x00000010
#define FCSR_Flag_U	0x00000008
#define FCSR_Flag_I	0x00000004
